clippy_lints/src/returns.rs

```diff
@@ -353,23 +369,68 @@ fn check_final_expr<'tcx>(
             emit_return_lint(cx, ret_span, semi_spans, &replacement);
         },
         ExprKind::If(_, then, else_clause_opt) => {
-            check_block_return(cx, &then.kind, peeled_drop_expr.span, semi_spans.clone());
+            check_block_return(
+                cx,
+                &then.kind,
+                peeled_drop_expr.span,
+                semi_spans.clone(),
+                parent_fn_return_type,
+            );
             if let Some(else_clause) = else_clause_opt {
-                check_block_return(cx, &else_clause.kind, peeled_drop_expr.span, semi_spans);
+                check_block_return(
+                    cx,
+                    &else_clause.kind,
+                    peeled_drop_expr.span,
+                    semi_spans,
+                    parent_fn_return_type,
+                );
             }
         },
         // a match expr, check all arms
         // an if/if let expr, check both exprs
         // note, if without else is going to be a type checking error anyways
         // (except for unit type functions) so we don't match it
-        ExprKind::Match(_, arms, MatchSource::Normal) => {
-            let match_ty = cx.typeck_results().expr_ty(peeled_drop_expr);
-            for arm in *arms {
-                check_final_expr(cx, arm.body, semi_spans.clone(), RetReplacement::Unit, Some(match_ty));
+        ExprKind::Match(match_expr, arms, MatchSource::Normal) => {
+            // need some special checks for cases where we would try to convert
+            // match statement to match expression
+            let check_arms = if parent_stmt_kind_opt.is_some_and(|x| matches!(x, StmtKind::Semi(_))) {
+                let match_expr_ret_type = cx.typeck_results().expr_ty(match_expr);
+
+                // check if function return type is equal to match return type
+                // if yes then we want can convert this match stmt to match expr
+                // otherwise dont do it
+                match parent_fn_return_type {
+                    Some(FnRetTy::DefaultReturn(_)) if let ty::Tuple(tuple) = match_expr_ret_type.kind() =>
+                        tuple.len() == 0,
+                    Some(FnRetTy::Return(_ret_ty)) => matches!(match_expr_ret_type.kind(), _ret_ty),
```

Comment:
Ah I was wondering why it didn't work, when I checked the type back then both showed `TyKind`, didn't realize they were from different module 🤦🏻 

So for the suggestion, I think that won't work out cause of this comment:
```
check if function return type is equal to match return type 
if yes then we want can convert this match stmt to match expr 
otherwise dont do it
```

We want to compare parent function's return type to match's type, and if they are the same then we can do the conversion of removing semicolon from end, etc.

Reply:
For non unit types that wouldn't type check since the implicit return of `()` would not apply, e.g. the following doesn't compile

```rust
fn f() -> u32 {
    match 1 {
        0 => 1, // ERROR: mismatched types
        _ => return 2,
    };
}
```

Reply:
Yeah I'll give it another shot, I had trouble trying to explain it so I don't blame you 😄

To annotate the example from #10390:

```rust
fn func(input: u8) {
    match input {
        // this arm's expr has type `i32`
        3 => 4,
        
        // this arm's expr has a pre-adjustment type of `!` (never) because of
        // the return, but adjusts to `i32` to match the other arm
        _ => {
            return;
        }, 
    };
    // the match expression has type i32, which is discarded by the semicolon statement

    // if the `3 => 4` arm was taken, the function returns () implicitly here
}
```

The FP occurs because of the combination of the discarded result and the implicit return of `()`

When a function isn't returning unit it can no longer use that implicit return because it would be a type mismatch, so every arm must diverge (return/panic/etc)

Reply:
Technically there is still a false positive this way, but it's already a false positive in the current implementation and isn't that likely to actually occur:

```rust
fn returns_str(input: u8) -> &'static str {
    match input {
        1 => return "a",
        _ => {
            return "b";
            // dead code, but would still be a type error if the ; is removed
            2
        },
    };
}
```

Reply:
Thanks a lot for taking another stab at explaining me this!

I tried to go through our discussion from start to re-understand if I am not missing something, I will try to replay it again here to make sure I am understanding everything properly:

- We started with discussing why patch's match and function return type match part of code will always return `true`, which made perfect sense, and next you suggested that we can directly convert it to just use unit types.
- Patch currently covers implicit returns case which is unit type again, but denoted as `FnRetTy::DefaultReturn` in code. But it would fail when passed an explicit unit return type of function, which is a perfect point. Your suggestion also asked if there was a case where non-unit types in explicit returns would fail.
- And then I said I want that to work for cases where semicolon and return needs to be removed for cases where match and function return type match. To which you replied that the case for:
```rust
fn f() -> u32 {
    match 1 {
        0 => 1, // ERROR: mismatched types
        _ => return 2,
    };
}
```
Wouldn't as such matter cause it does not compile anyways. Okay I think that makes sense till now. But one case I think we are missing is this:
```rust
fn f() -> u32 {
    return match 1 {
        0 => 1,
        _ => return 2,
    };
}
```
In this case too we want to suggest that please remove both returns and semicolon, this would only be possible if match's and function's return type match.

As in for the last case you mentioned, it anyways does not make sense to leave dead code around, other cases we are discussing here can and will happen in real code, so ig that's safe to ignore? 

I really hope this de-confuses both of us and I am able to give a proper answer to your initial question 😅

Reply:
> In this case too we want to suggest that please remove both returns and semicolon, this would only be possible if match's and function's return type match.

Since this example is returning the match expression the value of it is no longer discarded, the arms must match the return type of the function (or diverge). e.g. if you change the `1` to `"a"` it will be a type error

> As in for the last case you mentioned, it anyways does not make sense to leave dead code around, other cases we are discussing here can and will happen in real code, so ig that's safe to ignore?

Yeah it's fine to ignore I would say, it won't be that common and it's not a regression since it's already linted

Reply:
> this example is returning the match expression

Ah ig this was the point I was missing, here it is match expr 🤦🏻, yeah then I guess your stmt of just checking unit tuple makes sense, lemme make a fix and push it!

Reply:
Okay, so interestingly, we have another lint called `useless-unit`:
```
+++ <stderr output>
+error: unneeded unit return type
+  --> $DIR/needless_return.rs:336:52
+   |
+LL | fn issue10390_different_arm_return_types(input: u8) -> () {
+   |                                                    ^^^^^^ help: remove the `-> ()`
+   |
+   = note: `-D clippy::unused-unit` implied by `-D warnings`
```

Applying this would make it same as default case, do we still keep this part of code around then? I am not sure what is the policy of project in cases where another lint applies and hence changes output of it 😅

Reply:
Oh, I think I found a case where we may need to check types, what about this:

```rust
fn issue8156(x: u8) -> u64 {
    match x {
        80 => {
            return 10;
        },
        _ => {
            return 100;
        },
    };
}
```

In this case, it is a match statement.

Reply:
What would it be guarding against? That example compiles fine if you remove the return/semicolons